File: AndroidShimmerDemo/app/src/main/java/com/example/shakil/androidshimmerdemo/MainActivity.java
package com.example.shakil.androidshimmerdemo;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;

import com.example.shakil.androidshimmerdemo.Adapter.MyAdapter;
import com.example.shakil.androidshimmerdemo.Model.Model;
import com.example.shakil.androidshimmerdemo.Retrofit.IMyApi;
import com.example.shakil.androidshimmerdemo.Retrofit.RetrofitClient;

import java.util.List;

import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import io.supercharge.shimmerlayout.ShimmerLayout;
import retrofit2.Retrofit;

public class MainActivity extends AppCompatActivity {

    IMyApi myApi;
    RecyclerView recyclerView;
    CompositeDisposable compositeDisposable = new CompositeDisposable();
    ShimmerLayout shimmerLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //Init API
        Retrofit retrofit = RetrofitClient.getInstance();
        myApi = retrofit.create(IMyApi.class);

        shimmerLayout = findViewById(R.id.shimmerLayout);

        recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        //Fetch data
        fetchDate();
    }

    private void fetchDate() {
        shimmerLayout.startShimmerAnimation();

        compositeDisposable.add(myApi.getDate()
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(new Consumer<List<Model>>() {
            @Override
            public void accept(List<Model> models) throws Exception {
                displayData(models);
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(Throwable throwable) throws Exception {
                Toast.makeText(MainActivity.this, throwable.getMessage(), Toast.LENGTH_SHORT).show();
            }
        }));
    }

    private void displayData(List<Model> models) {
        MyAdapter adapter = new MyAdapter(this, models);
        recyclerView.setAdapter(adapter);
        shimmerLayout.stopShimmerAnimation();
        shimmerLayout.setVisibility(View.GONE);
    }

    @Override
    protected void onStop() {
        compositeDisposable.clear();
        super.onStop();
    }
}

File: KotlinShimmerDemo/app/src/main/java/com/example/shakil/kotlinshimmerdemo/MainActivity.kt
package com.example.shakil.kotlinshimmerdemo

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.example.shakil.kotlinshimmerdemo.Adapter.MyAdapter
import com.example.shakil.kotlinshimmerdemo.Model.Model
import com.example.shakil.kotlinshimmerdemo.Retrofit.IMyAPI
import com.example.shakil.kotlinshimmerdemo.Retrofit.RetrofitClient
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    lateinit var myAPI: IMyAPI
    var compositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Init API
        val retrofit = RetrofitClient.instance
        myAPI = retrofit.create(IMyAPI::class.java)

        recycler_view.setHasFixedSize(true)
        recycler_view.layoutManager = LinearLayoutManager(this)

        fetchData()
    }

    private fun fetchData() {
        shimmerLayout.startShimmerAnimation()
        compositeDisposable.add(myAPI.data.subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { photos -> displayData(photos) })
    }

    private fun displayData(photos: List<Model>?) {
        val adapter = MyAdapter(this, photos!!)
        recycler_view.adapter = adapter
        shimmerLayout.stopShimmerAnimation()
        shimmerLayout.visibility = View.GONE
    }

    override fun onStop() {
        compositeDisposable.clear()
        super.onStop()
    }
}
